component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/voiceotp/VoiceOTPAuthenticator.java

```diff
@@ -0,0 +1,2037 @@
+import static org.wso2.carbon.identity.authenticator.voiceotp.VoiceOTPConstants.POST_METHOD;
+import static org.wso2.carbon.identity.authenticator.voiceotp.VoiceOTPConstants.REQUESTED_USER_MOBILE;
+import static org.wso2.carbon.identity.authenticator.voiceotp.VoiceOTPConstants.RESEND;
+import static org.wso2.carbon.identity.authenticator.voiceotp.VoiceOTPConstants.XML_CONTENT_TYPE;
+
+import static java.util.Base64.getEncoder;
+
+/**
+ * Authenticator of Voice OTP.
+ */
+public class VoiceOTPAuthenticator extends AbstractApplicationAuthenticator
+        implements FederatedApplicationAuthenticator {
+
+    private static final Log log = LogFactory.getLog(VoiceOTPAuthenticator.class);
+
+    /**
+     * Check whether the authentication or logout request can be handled by the authenticator.
+     *
+     * @param request The HttpServletRequest
+     * @return whether the authenticator can handle the inbound request
+     */
+    @Override
+    public boolean canHandle(HttpServletRequest request) {
+
+        log.debug("Inside VoiceOTPAuthenticator canHandle method and check the existence of mobile number and " +
+                    "otp code");
+
+        return ((StringUtils.isNotEmpty(request.getParameter(VoiceOTPConstants.RESEND))
+                && StringUtils.isEmpty(request.getParameter(VoiceOTPConstants.CODE)))
+                || StringUtils.isNotEmpty(request.getParameter(VoiceOTPConstants.CODE))
+                || StringUtils.isNotEmpty(request.getParameter(VoiceOTPConstants.MOBILE_NUMBER)));
+    }
+
+    /**
+     * Check whether the authentication or logout request can be handled by the authenticator.
+     *
+     * @param request The HttpServletRequest
+     * @param response The HttpServletResponse
+     * @param context The AuthenticationContext
+     * @return AuthenticatorFlowStatus type
+     */
+    @Override
+    public AuthenticatorFlowStatus process(HttpServletRequest request,
+                                           HttpServletResponse response,
+                                           AuthenticationContext context)
+            throws AuthenticationFailedException, LogoutFailedException {
+
+        // If the logout request comes, then no need to go through and complete the flow.
+        if (context.isLogoutRequest()) {
+            return AuthenticatorFlowStatus.SUCCESS_COMPLETED;
+        } else if (StringUtils.isNotEmpty(request.getParameter(VoiceOTPConstants.MOBILE_NUMBER))) {
+            // if the request comes with MOBILE_NUMBER, it will go through this flow.
+            initiateAuthenticationRequest(request, response, context);
+            return AuthenticatorFlowStatus.INCOMPLETE;
+        } else if (StringUtils.isEmpty(request.getParameter(VoiceOTPConstants.CODE))) {
+            // if the request comes with code, it will go through this flow.
+            initiateAuthenticationRequest(request, response, context);
+            if (context.getProperty(VoiceOTPConstants.AUTHENTICATION)
```

Comment:
In the initiateAuthenticationRequest we're setting the authentication property to the context. Is this check necessary ?

Reply:
This is similer to the logic in totp and smsotp authenticators.

https://github.com/madurangasiriwardena/identity-outbound-auth-totp/blob/2945758e8fa1b788a5e1af8c78f704eea92b9a0e/component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java#L108

https://github.com/senthalan/identity-outbound-auth-sms-otp/blob/86fe41f492b509f30644535294b42155666d3bb9/component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java#L117